<a href="https://colab.research.google.com/github/BEMiklos/PlantTraits2024/blob/main/XGB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install -q kaggle

In [4]:
!mkdir ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

mkdir: cannot create directory ‘/root/.kaggle’: File exists


In [5]:
!kaggle competitions download -c planttraits2024

 99% 3.15G/3.17G [00:27<00:00, 301MB/s]
100% 3.17G/3.17G [00:27<00:00, 126MB/s]


In [6]:
!unzip -q '/content/planttraits2024.zip'

In [67]:
import random

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import albumentations as A
import cv2

import torch
from torch import nn
import torch.nn.functional as F
from torchvision.io import read_image
from torch.utils.data import Dataset, ConcatDataset, DataLoader, Subset, random_split
from torchvision import transforms

from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.multioutput import MultiOutputRegressor
from sklearn.metrics import r2_score
from sklearn.preprocessing import PolynomialFeatures

import xgboost as xgb

import re

In [2]:
device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
random_state = 7

np.random.seed(0)

random.seed(0)

torch.manual_seed(0)
torch.cuda.manual_seed(0)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

In [3]:
TRAIN_IMAGES_PATH = '/content/train_images'
TEST_IMAGES_PATH = '/content/test_images'

### Untility functions

In [4]:
mean_pattern = r'X\d+_mean'
sd_pattern = r'X\d+_sd'

In [5]:
def find_matching_elements(lst,pattern):
    matching_elements = [element for element in lst if re.search(pattern, element)]
    return matching_elements

In [6]:
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

In [7]:
train_df.sample(10)

id  WORLDCLIM_BIO1_annual_mean_temperature  \
24886  184873835                               -1.032917   
31965  190494649                                0.440794   
28421  195642066                               17.284445   
12674  164426751                               12.075297   
31158  186882920                               17.013542   
45443  195637432                               -0.526925   
19727  194871319                               20.958969   
20600  186386731                               18.402203   
24052  164361896                               17.402231   
35949  195065314                               11.712444   

       WORLDCLIM_BIO12_annual_precipitation  \
24886                            574.295227   
31965                            589.538086   
28421                            341.909515   
12674                            166.142853   
31158                           1030.795898   
45443                            317.861908   
19727                           2068.119141   
20600                            563.590454   
24052                            870.571411   
35949                            437.182220   

       WORLDCLIM_BIO13.BIO14_delta_precipitation_of_wettest_and_dryest_month  \
24886                                          90.252380                       
31965                                          48.000000                       
28421                                          65.457146                       
12674                                          24.500000                       
31158                                         266.556122                       
45443                                          87.438095                       
19727                                         264.828583                       
20600                                          81.157143                       
24052                                          82.178574                       
35949                                          34.533333                       

       WORLDCLIM_BIO15_precipitation_seasonality  \
24886                                  62.482738   
31965                                  33.024284   
28421                                  89.490135   
12674                                  49.969826   
31158                                  91.027039   
45443                                 113.449104   
19727                                  53.143101   
20600                                  64.466095   
24052                                  35.694508   
35949                                  28.756390   

       WORLDCLIM_BIO4_temperature_seasonality  \
24886                             1227.480225   
31965                             1090.136475   
28421                              544.396545   
12674                              902.658264   
31158                              177.292847   
45443                             1096.622192   
19727                              488.990204   
20600                              437.827026   
24052                              825.165405   
35949                              632.724487   

       WORLDCLIM_BIO7_temperature_annual_range  SOIL_bdod_0.5cm_mean_0.01_deg  \
24886                                42.762859                             87   
31965                                39.804287                             54   
28421                                31.776667                            138   
12674                                38.064285                            129   
31158                                14.822959                            107   
45443                                40.898571                            108   
19727                                19.647144                            115   
20600                                24.329046                            121   
24052                                35.850002                            148   
35949                          

In [8]:
len(train_df.columns), len(test_df.columns)

(176, 164)

In [9]:
len(train_df), len(test_df)

(55489, 6545)

In [10]:
targets = [col for col in train_df.columns if col not in test_df.columns]
targets = targets[:6]

In [11]:
metadata_columns = [col for col in train_df.columns if col in test_df.columns and col != 'id']

 TODO: hiányzó értékek kezelésének pontosítása

In [12]:
# Check for missing values in the train DataFrame
train_missing_values = train_df.isnull().sum()
print("Missing values in Train DataFrame:")
print(train_missing_values)

Missing values in Train DataFrame:
id                                                                           0
WORLDCLIM_BIO1_annual_mean_temperature                                       0
WORLDCLIM_BIO12_annual_precipitation                                         0
WORLDCLIM_BIO13.BIO14_delta_precipitation_of_wettest_and_dryest_month        0
WORLDCLIM_BIO15_precipitation_seasonality                                    0
                                                                         ...  
X11_sd                                                                   16341
X18_sd                                                                   16341
X26_sd                                                                   16341
X50_sd                                                                   16341
X3112_sd                                                                 16341
Length: 176, dtype: int64


In [13]:
train_df.fillna(train_df.mean(), inplace=True)

In [14]:
columns_to_log = ['X11_mean', 'X18_mean', 'X26_mean', 'X50_mean', 'X3112_mean',
                  'X4_sd', 'X11_sd', 'X18_sd', 'X26_sd', 'X50_sd', 'X3112_sd']

# Apply log10 transformation to the specified columns
for column in columns_to_log:
    train_df[column] = np.log10(train_df[column] + 1)  # Adding 1 to avoid log(0)


In [15]:
def split_dataframe(train_df, test_df, train_size=0.8, val_size=0.2):
    if train_size + val_size != 1:
        raise ValueError("train_size, val_size, and test_size must sum to 1")

    train_df, val_df = train_test_split(train_df, train_size=train_size, random_state=random_state)
    return train_df, val_df, test_df

In [16]:
def r2_loss(output, target):
    target = target
    output = output
    target_mean = torch.mean(target)
    ss_tot = torch.sum((target - target_mean) ** 2)
    ss_res = torch.sum((target - output) ** 2)
    r2 = 1 - ss_res / ss_tot
    return 1 - r2

In [24]:
!pip -q install torchmetrics

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 841.5/841.5 kB 4.9 MB/s eta 0:00:00


In [32]:
from torchmetrics.regression import R2Score

# Data preparation

In [19]:
bemeno_valtozok = metadata_columns

In [20]:
celvaltozok = targets[:6]
print("Targets:",celvaltozok)

Targets: ['X4_mean', 'X11_mean', 'X18_mean', 'X26_mean', 'X50_mean', 'X3112_mean']


In [21]:
train_df, val_df, test_df = split_dataframe(train_df.sample(frac=1, random_state=1), test_df.sample(frac=1, random_state=1))

In [22]:
train_df

id  WORLDCLIM_BIO1_annual_mean_temperature  \
24239  171376117                               15.120000   
26870  196019212                               13.562074   
40848  173458431                               26.131250   
51793  176675400                                2.116071   
21466   74535495                               21.885370   
...          ...                                     ...   
25516  188291299                               12.671389   
47484  141070257                               19.723888   
20133  195683798                               10.255437   
39678  194874328                               19.927977   
21538  170649825                               15.096488   

       WORLDCLIM_BIO12_annual_precipitation  \
24239                           1123.524414   
26870                           1100.364502   
40848                           3183.500000   
51793                            887.904785   
21466                            331.084442   
...                                     ...   
25516                            816.938110   
47484                            114.900002   
20133                           1059.966675   
39678                           3078.928467   
21538                           1196.038086   

       WORLDCLIM_BIO13.BIO14_delta_precipitation_of_wettest_and_dryest_month  \
24239                                         179.373337                       
26870                                         180.542221                       
40848                                         528.928589                       
51793                                         151.476196                       
21466                                          64.617775                       
...                                                  ...                       
25516                                         141.033340                       
47484                                          34.566666                       
20133                                          45.538094                       
39678                                         269.892853                       
21538                                         191.138092                       

       WORLDCLIM_BIO15_precipitation_seasonality  \
24239                                  75.942123   
26870                                  73.934456   
40848                                  65.707703   
51793                                  68.991066   
21466                                  90.718796   
...                                          ...   
25516                                  72.865730   
47484                                 114.762772   
20133                                  16.683119   
39678                                  38.449482   
21538                                  75.633522   

       WORLDCLIM_BIO4_temperature_seasonality  \
24239                              307.151367   
26870                              289.352539   
40848                               76.344696   
51793                               99.114906   
21466                              418.202576   
...                                       ...   
25516                              190.145630   
47484                              240.889420   
20133                              899.692749   
39678                              467.681274   
21538                              275.434082   

       WORLDCLIM_BIO7_temperature_annual_range  SOIL_bdod_0.5cm_mean_0.01_deg  \
24239                                19.777777                            117   
26870                                19.765333                            120   
40848                                12.764285                             95   
51793                                16.292856                            112   
21466                                25.793777                            153   
...                                        ...                     

# Feature engineering

In [36]:
worldclim = train_df.columns[1:7]
soil = train_df.columns[7:68]
modis = train_df.columns[68:128]
vod = train_df.columns[128:164]

In [37]:
targets = ['X4_mean', 'X11_mean', 'X18_mean', 'X26_mean', 'X50_mean','X3112_mean', 'X4_sd',
 'X11_sd',
 'X18_sd',
 'X26_sd',
 'X50_sd',
 'X3112_sd']

In [38]:
feature_cols = [col for col in train_df.columns if col not in targets]
features = train_df[feature_cols]

In [46]:
targets = ['X4_mean', 'X11_mean', 'X18_mean', 'X26_mean', 'X50_mean','X3112_mean']

In [39]:
imputer = SimpleImputer(strategy='median')

features_imputed = imputer.fit_transform(features)
features_imputed = pd.DataFrame(features_imputed, columns=features.columns)

In [40]:
scaler = StandardScaler()

features_scaled = scaler.fit_transform(features_imputed)
features_scaled = pd.DataFrame(features_scaled, columns=features.columns)

In [47]:
y_multi = train_df[targets]
X_train_multi, X_test_multi, y_train_multi, y_test_multi = train_test_split(
    features_scaled[feature_cols], y_multi, test_size=0.2, random_state=42)

multi_xgb_model = MultiOutputRegressor(xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42))

multi_xgb_model.fit(X_train_multi, y_train_multi)

y_pred_multi = multi_xgb_model.predict(X_test_multi)

r2_scores = r2_score(y_test_multi, y_pred_multi, multioutput='raw_values')
print("R² scores for each target (on scaled features):", r2_scores)

R² scores for each target (on scaled features): [ 0.06519863  0.16375705  0.3071952   0.23258292 -0.24942554  0.25492991]


## Dimension reduction based on importance

In [48]:
X_train_full, X_test_full, y_train_full, y_test_full = train_test_split(features_scaled, train_df[targets], test_size=0.2, random_state=42)

xgb_model_full = xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42)
xgb_model_full.fit(X_train_full, y_train_full)

importances = xgb_model_full.feature_importances_

feature_names = features_scaled.columns
importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': importances})
importance_df = importance_df.sort_values(by='Importance', ascending=False).reset_index(drop=True)
print(importance_df.head(20))

                                              Feature  Importance
0              WORLDCLIM_BIO1_annual_mean_temperature    0.053923
1             WORLDCLIM_BIO7_temperature_annual_range    0.032605
2              WORLDCLIM_BIO4_temperature_seasonality    0.019557
3   MODIS_2000.2020_monthly_mean_surface_reflectan...    0.015389
4                     SOIL_ocd_60.100cm_mean_0.01_deg    0.010839
5   MODIS_2000.2020_monthly_mean_surface_reflectan...    0.009633
6                    SOIL_sand_60.100cm_mean_0.01_deg    0.009385
7                WORLDCLIM_BIO12_annual_precipitation    0.008858
8   WORLDCLIM_BIO13.BIO14_delta_precipitation_of_w...    0.008784
9   MODIS_2000.2020_monthly_mean_surface_reflectan...    0.008703
10                   SOIL_phh2o_30.60cm_mean_0.01_deg    0.008532
11                VOD_Ku_1987_2017_multiyear_mean_m08    0.008341
12                    SOIL_silt_30.60cm_mean_0.01_deg    0.008260
13                 VOD_X_1997_2018_multiyear_mean_m10    0.008218
14        

In [49]:
top_features = importance_df['Feature'].head(30).tolist()
y_multi = train_df[targets]
X_train_multi, X_test_multi, y_train_multi, y_test_multi = train_test_split(
    features_scaled[top_features], y_multi, test_size=0.2, random_state=42)

multi_xgb_model = MultiOutputRegressor(xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42))

multi_xgb_model.fit(X_train_multi, y_train_multi)

y_pred_multi = multi_xgb_model.predict(X_test_multi)

r2_scores = r2_score(y_test_multi, y_pred_multi, multioutput='raw_values')
print("R² scores for each target (with the top 30 important features):", r2_scores)

R² scores for each target (with the top 30 important features): [ 0.093593    0.15704444  0.303049    0.228631   -0.18055592  0.24359068]


R² scores for each target (on scaled features): [ 0.06519863,  0.16375705,  0.3071952,   0.23258292, -0.24942554,  0.25492991]

R² scores for each target (with the top 30 important features): [ 0.093593,    0.15704444,  0.303049,    0.228631,   -0.18055592,  0.24359068]

# Dimension reduction with PCA

In [50]:
pca = PCA(n_components=min(features_scaled.shape))

pca.fit(features_scaled)

cumulative_variance_ratio = np.cumsum(pca.explained_variance_ratio_)

In [51]:
n_components_95 = np.where(cumulative_variance_ratio >= 0.95)[0][0] + 1

pca_95 = PCA(n_components=n_components_95)

features_pca_95 = pca_95.fit_transform(features_scaled)

features_pca_95_df = pd.DataFrame(features_pca_95, columns=[f'PC{i+1}' for i in range(n_components_95)])
features_pca_95_df.head(), n_components_95

(        PC1       PC2       PC3       PC4       PC5       PC6       PC7  \
 0  1.752592 -4.705826 -1.658646 -0.237254  0.793511 -2.338007 -1.751561   
 1  2.015580 -2.177639 -2.931064  0.877646  2.630513 -0.920006 -1.026357   
 2  6.143819 -6.758618  4.730867  0.514096 -1.498595  0.882158 -4.123164   
 3  2.972285  1.678727  2.701243 -2.987341 -0.353421  4.852251 -1.072083   
 4 -7.082553 -2.660474 -5.109291 -4.465095  1.849223 -1.732129 -0.115992   
 
         PC8       PC9      PC10      PC11      PC12      PC13      PC14  \
 0  2.650341  0.879054  0.370322 -1.133138 -0.110596 -0.455378  0.193836   
 1  0.542654  1.081605 -0.052364 -0.253646 -0.305663 -1.651285  0.337962   
 2  5.063336 -0.348512  3.381313 -1.122087  1.348846  0.416042 -2.509848   
 3 -0.430468 -3.790803  1.163000 -1.283125 -0.621630 -0.566913  0.409377   
 4 -0.483845  1.874738  0.996085 -0.396926 -0.226993 -0.830693  1.196308   
 
        PC15      PC16      PC17      PC18      PC19  
 0 -1.147980 -0.223689 -0.908

In [54]:
X_train_multi, X_test_multi, y_train_multi, y_test_multi = train_test_split(
    features_pca_95_df, y_multi, test_size=0.2, random_state=42)

multi_xgb_model = MultiOutputRegressor(xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42))

multi_xgb_model.fit(X_train_multi, y_train_multi)

y_pred_multi = multi_xgb_model.predict(X_test_multi)

r2_scores = r2_score(y_test_multi, y_pred_multi, multioutput='raw_values')
print("R² scores for each target (PCA):", r2_scores)

R² scores for each target (PCA): [ 0.0397597   0.15494675  0.28841653  0.21596206 -0.2079703   0.24695846]


R² scores for each target (on scaled features): [ 0.06519863, 0.16375705, 0.3071952, 0.23258292, -0.24942554, 0.25492991]

R² scores for each target (with the top 30 important features): [ 0.093593, 0.15704444, 0.303049, 0.228631, -0.18055592, 0.24359068]

R² scores for each target (PCA): [ 0.0397597,   0.15494675,  0.28841653,  0.21596206, -0.2079703,   0.24695846]

In [65]:
X_train_full, X_test_full, y_train_full, y_test_full = train_test_split(features_pca_95_df, train_df[targets], test_size=0.2, random_state=42)

xgb_model_full = xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42)
xgb_model_full.fit(X_train_full, y_train_full)

importances = xgb_model_full.feature_importances_

feature_names = features_pca_95_df.columns
importance_df_pca = pd.DataFrame({'Feature': feature_names, 'Importance': importances})
importance_df_pca = importance_df.sort_values(by='Importance', ascending=False).reset_index(drop=True)
print(importance_df_pca.head(20))

                                              Feature  Importance
0              WORLDCLIM_BIO1_annual_mean_temperature    0.053923
1             WORLDCLIM_BIO7_temperature_annual_range    0.032605
2              WORLDCLIM_BIO4_temperature_seasonality    0.019557
3   MODIS_2000.2020_monthly_mean_surface_reflectan...    0.015389
4                     SOIL_ocd_60.100cm_mean_0.01_deg    0.010839
5   MODIS_2000.2020_monthly_mean_surface_reflectan...    0.009633
6                    SOIL_sand_60.100cm_mean_0.01_deg    0.009385
7                WORLDCLIM_BIO12_annual_precipitation    0.008858
8   WORLDCLIM_BIO13.BIO14_delta_precipitation_of_w...    0.008784
9   MODIS_2000.2020_monthly_mean_surface_reflectan...    0.008703
10                   SOIL_phh2o_30.60cm_mean_0.01_deg    0.008532
11                VOD_Ku_1987_2017_multiyear_mean_m08    0.008341
12                    SOIL_silt_30.60cm_mean_0.01_deg    0.008260
13                 VOD_X_1997_2018_multiyear_mean_m10    0.008218
14        

In [66]:
top_features = importance_df_pca['Feature'].head(30).tolist()
y_multi = train_df[targets]
X_train_multi, X_test_multi, y_train_multi, y_test_multi = train_test_split(
    features_scaled[top_features], y_multi, test_size=0.2, random_state=42)

multi_xgb_model = MultiOutputRegressor(xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42))

multi_xgb_model.fit(X_train_multi, y_train_multi)

y_pred_multi = multi_xgb_model.predict(X_test_multi)

r2_scores = r2_score(y_test_multi, y_pred_multi, multioutput='raw_values')
print("R² scores for each target (PCA with the top 30 important features):", r2_scores)

R² scores for each target (PCA with the top 30 important features): [ 0.093593    0.15704444  0.303049    0.228631   -0.18055592  0.24359068]


R² scores for each target (on scaled features): [ 0.06519863, 0.16375705, 0.3071952, 0.23258292, -0.24942554, 0.25492991]

R² scores for each target (with the top 30 important features): [ 0.093593, 0.15704444, 0.303049, 0.228631, -0.18055592, 0.24359068]

R² scores for each target (PCA): [ 0.0397597, 0.15494675, 0.28841653, 0.21596206, -0.2079703, 0.24695846]

R² scores for each target (PCA with the top 30 important features): [ 0.093593,    0.15704444,  0.303049,    0.228631,   -0.18055592,  0.24359068]

In [74]:
from sklearn.model_selection import GridSearchCV

In [75]:
top_features = importance_df_pca['Feature'].head(30).tolist()
targets = ['X4_mean', 'X11_mean', 'X18_mean', 'X26_mean', 'X50_mean','X3112_mean']
y_multi = train_df[targets]
X = features_scaled[top_features]

X_train_multi, X_test_multi, y_train_multi, y_test_multi = train_test_split(
    X, y_multi, test_size=0.2, random_state=42)

xgb_base = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)

multi_xgb = MultiOutputRegressor(xgb_base)

param_grid = {
    'estimator__n_estimators': [50, 100, 200],
    'estimator__max_depth': [3, 5, 7],
    'estimator__learning_rate': [0.01, 0.1, 0.2]
}

grid_search = GridSearchCV(estimator=multi_xgb, param_grid=param_grid,
                           cv=3, n_jobs=-1, verbose=2, scoring='r2')

grid_search.fit(X_train_multi, y_train_multi)

print("Best parameters found: ", grid_search.best_params_)
print("Best R² score found: ", grid_search.best_score_)

Fitting 3 folds for each of 27 candidates, totalling 81 fits


/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/fork_exec.py:38: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid = os.fork()


Best parameters found:  {'estimator__learning_rate': 0.1, 'estimator__max_depth': 5, 'estimator__n_estimators': 100}
Best R² score found:  0.1551971370767798


In [77]:
top_features = importance_df_pca['Feature'].head(30).tolist()
targets = ['X4_mean', 'X11_mean', 'X18_mean', 'X26_mean', 'X50_mean','X3112_mean']
y_multi = train_df[targets]
X_train_multi, X_test_multi, y_train_multi, y_test_multi = train_test_split(
    features_scaled[top_features], y_multi, test_size=0.2, random_state=42)

multi_xgb_model = MultiOutputRegressor(xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=42, learning_rate=0.1, max_depth=5))

multi_xgb_model.fit(X_train_multi, y_train_multi)

y_pred_multi = multi_xgb_model.predict(X_test_multi)

r2_scores = r2_score(y_test_multi, y_pred_multi, multioutput='raw_values')
print("R² scores for each target (PCA with the top 30 important features):", r2_scores)

R² scores for each target (PCA with the top 30 important features): [0.12358511 0.17924043 0.31857293 0.24550138 0.01399558 0.26230538]


R² scores for each target (PCA with the top 30 important features optimized): [0.12358511, 0.17924043, 0.31857293, 0.24550138, 0.01399558, 0.26230538]

 R² scores for each target (PCA with the top 30 important features): [ 0.093593, 0.15704444, 0.303049, 0.228631, -0.18055592, 0.24359068]

